From e9cc437113f230117d4515762c9c93841dc16f39 Mon Sep 17 00:00:00 2001
From: retro98boy <retro98boy@qq.com>
Date: Sat, 25 Jul 2026 19:17:13 +0800
Subject: [PATCH 02/20] arm64: dts: rockchip: add USB3.0 and USB2.0 nodes for
 RK3562

After testing, this [patch](https://github.com/armbian/build/blob/90beab3d2e713d9aac5b1f8c272fc4f92b849dfc/patch/kernel/archive/rockchip64-7.1/rk3528-10-phy-rockchip-inno-usb2-Add-support-for-RK3528.patch) is also required for the USB PHY to probe successfully. Otherwise, the following error occurs:

```bash
[   10.984491] platform ff740000.usb2-phy: deferred probe pending: (reason unknown)
[   10.985234] platform fe500000.usb: deferred probe pending: platform: wait for supplier /soc/usb2-phy@ff740000/otg-port
[   10.986297] platform fed40000.usb: deferred probe pending: platform: wait for supplier /soc/usb2-phy@ff740000/host-port
```
---
 arch/arm64/boot/dts/rockchip/rk3562.dtsi | 81 ++++++++++++++++++++++++
 1 file changed, 81 insertions(+)

diff --git a/arch/arm64/boot/dts/rockchip/rk3562.dtsi b/arch/arm64/boot/dts/rockchip/rk3562.dtsi
index 21f604579..47d395b0e 100644
--- a/arch/arm64/boot/dts/rockchip/rk3562.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3562.dtsi
@@ -371,6 +371,40 @@ pcie2x1_intc: legacy-interrupt-controller {
 			};
 		};
 
+		usbdrd3: usb@fe500000 {
+			compatible = "rockchip,rk3562-dwc3", "rockchip,rk3399-dwc3";
+			clocks = <&cru CLK_USB3OTG_REF>, <&cru CLK_USB3OTG_SUSPEND>,
+				 <&cru ACLK_USB3OTG>, <&cru PCLK_PHP>;
+			clock-names = "ref", "suspend", "bus", "pipe_clk";
+			resets = <&cru SRST_USB3OTG>;
+			reset-names = "usb3-otg";
+			#address-cells = <2>;
+			#size-cells = <2>;
+			ranges;
+			status = "disabled";
+
+			usbdrd_dwc3: usb@fe500000 {
+				compatible = "snps,dwc3";
+				reg = <0x0 0xfe500000 0x0 0x400000>;
+				interrupts = <GIC_SPI 149 IRQ_TYPE_LEVEL_HIGH>;
+				dr_mode = "otg";
+				phys = <&u2phy_otg>, <&combphy PHY_TYPE_USB3>;
+				phy-names = "usb2-phy", "usb3-phy";
+				phy_type = "utmi_wide";
+				power-domains = <&power RK3562_PD_PHP>;
+				snps,dis_enblslpm_quirk;
+				/* snps,dis-u1-entry-quirk; */
+				/* snps,dis-u2-entry-quirk; */
+				snps,dis-u2-freeclk-exists-quirk;
+				snps,dis-del-phy-power-chg-quirk;
+				snps,dis-tx-ipgap-linecheck-quirk;
+				snps,dis_rxdet_inp3_quirk;
+				/* snps,parkmode-disable-hs-quirk; */
+				/* snps,parkmode-disable-ss-quirk; */
+				status = "disabled";
+			};
+		};
+
 		gic: interrupt-controller@fe901000 {
 			compatible = "arm,gic-400";
 			#interrupt-cells = <3>;
@@ -383,6 +417,26 @@ gic: interrupt-controller@fe901000 {
 			interrupts = <GIC_PPI 9 (GIC_CPU_MASK_SIMPLE(4) | IRQ_TYPE_LEVEL_LOW)>;
 		};
 
+		usb_host_ehci: usb@fed00000 {
+			compatible = "generic-ehci";
+			reg = <0x0 0xfed00000 0x0 0x40000>;
+			interrupts = <GIC_SPI 150 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cru HCLK_USB2HOST>, <&cru HCLK_USB2HOST_ARB>;
+			phys = <&u2phy_host>;
+			phy-names = "usb2-phy";
+			status = "disabled";
+		};
+
+		usb_host_ohci: usb@fed40000 {
+			compatible = "generic-ohci";
+			reg = <0x0 0xfed40000 0x0 0x40000>;
+			interrupts = <GIC_SPI 151 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cru HCLK_USB2HOST>, <&cru HCLK_USB2HOST_ARB>;
+			phys = <&u2phy_host>;
+			phy-names = "usb2-phy";
+			status = "disabled";
+		};
+
 		qos_dma2ddr: qos@fee03800 {
 			compatible = "rockchip,rk3562-qos", "syscon";
 			reg = <0x0 0xfee03800 0x0 0x20>;
@@ -1018,6 +1072,33 @@ saradc0: adc@ff730000 {
 			status = "disabled";
 		};
 
+		u2phy: usb2-phy@ff740000 {
+			compatible = "rockchip,rk3562-usb2phy";
+			reg = <0x0 0xff740000 0x0 0x10000>;
+			clocks = <&cru CLK_USB2PHY_REF>, <&cru PCLK_USB2PHY>;
+			clock-names = "phyclk", "pclk";
+			#clock-cells = <0>;
+			clock-output-names = "usb480m_phy";
+			rockchip,usbgrf = <&usbphy_grf>;
+			status = "disabled";
+
+			u2phy_otg: otg-port {
+				#phy-cells = <0>;
+				interrupts = <GIC_SPI 153 IRQ_TYPE_LEVEL_HIGH>,
+					     <GIC_SPI 154 IRQ_TYPE_LEVEL_HIGH>,
+					     <GIC_SPI 155 IRQ_TYPE_LEVEL_HIGH>;
+				interrupt-names = "otg-bvalid", "otg-id", "linestate";
+				status = "disabled";
+			};
+
+			u2phy_host: host-port {
+				#phy-cells = <0>;
+				interrupts = <GIC_SPI 156 IRQ_TYPE_LEVEL_HIGH>;
+				interrupt-names = "linestate";
+				status = "disabled";
+			};
+		};
+
 		combphy: phy@ff750000 {
 			compatible = "rockchip,rk3562-naneng-combphy";
 			reg = <0x0 0xff750000 0x0 0x100>;
-- 
2.55.0

